_posts/2021-01-27-#20833.md

```diff
@@ -0,0 +1,140 @@
+`CoinsViewBacked.GetCoin()` cannot find a coin in its own data structures, it may call its
+`Backend.GetCoin()`.
+
+- Since the UTXO set is quite large and stored on disk, Bitcoin Core uses a cache to improve
+validation speed.
+To prevent cache thrashing caused by an attacker broadcasting many invalid transactions spending a
+wide variety of UTXOs (a cheap DoS attack), we keep track of which coins are brought in from disk
+and uncache them if the transaction turns out to be invalid.
+
+- Validation roughly consists of `PreChecks`, `PolicyScriptChecks`, `ConsensusScriptChecks`, and
+`Finalize`.
+We've touched on these functions in previous review clubs, e.g.
+[#19339](https://bitcoincore.reviews/19339).
+[This Stack Exchange
+post](https://bitcoin.stackexchange.com/questions/100317/what-is-the-difference-between-policy-and-consensus-when-it-comes-to-a-bitcoin-c) about consensus vs policy rules may also be helpful.
+
+## Questions
+
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+
+2. What are some use cases for enabling packages through `testmempoolaccept`?
+
+3. From a code niceness standpoint, what is the benefit of moving the validation state, replaced
+transactions, and fee to `Workspace`?
+Should we add `coins_to_uncache` to `Workspace` and make `ATMPArgs` a const? Why or why not?
+
+4. Why do we run all `PreChecks` before script checks? Would it still be correct if we didn't do
+this?
+
+5. In
+[AcceptSingleTransaction](https://github.com/bitcoin/bitcoin/blob/574758e605e83d0219fbdb80a1bed4bf4ec71e2a/src/validation.cpp#L1037) (with `test_accept=true`), both `PolicyScriptChecks` and
+`ConsensusScriptChecks` are run.
+However, in
+[AcceptMultipleTransactions](https://github.com/bitcoin/bitcoin/blob/574758e605e83d0219fbdb80a1bed4bf4ec71e2a/src/validation.cpp#L1068), we don't run `ConsensusScriptChecks`. Why is that okay?
+(Followup question: why do we run `ConsensusScriptChecks` at all in ATMP?)
+
+6. When we're fetching a coin from `m_viewmempool`, what 4 places to we look and in what order? Why
+do we check `cache_package_remove` before `cache_package_add` in `CCoinsViewMemPool::GetCoin`?.
+
+7. If a block arrives while we're running `PreChecks`, do we have to worry about our `CoinsView`
+becoming inconsistent?
+
+8. Why don't we need to clear the `cache_package_add` and `cache_package_remove` in
+`CCoinsViewMemPool` between multiple calls to `ProcessNewPackage`?
+
+9. What do you think of commit
+[`e0a8e1f`](https://github.com/bitcoin/bitcoin/pull/20833/commits/e0a8e1fa9015eea685cb63e68ad1ec3da893fcd9), which removes the line switching the `m_view` backend to `m_dummy`?
+What kind of bug is it trying to prevent and how else are we trying to prevent such a bug?
+
+10. What are your thoughts on the results API (array of results for each tx, first one if failed)?
```

Comment:
```suggestion
10. What are your thoughts on the proposed `results` object in the RPC output (an array of results for each transaction, or the first one if failed)?
```

Reply:
Yeah it's weird because I'm proposing something to not change the current API 😅